ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoContractsSpec.scala

```diff
@@ -45,12 +46,114 @@ private[dao] trait JdbcLedgerDaoContractsSpec extends LoneElement with Inside {
     } yield {
       result shouldEqual None
     }
   }
 
+  it should "not find contracts that are not visible to any of the requesters" in {
+    for {
+      (_, tx) <- createAndStoreContract(
+        submittingParties = Set(alice),
+        signatories = Set(alice, bob),
+        stakeholders = Set(alice, bob),
+        key = None,
+      )
+      contractId = nonTransient(tx).loneElement
+      result <- ledgerDao.lookupActiveOrDivulgedContract(contractId, Set(charlie, emma))
+    } yield {
+      result shouldBe None
+    }
+  }
+
+  it should "find contract if at least one of requesters is a stakeholder" in {
+    for {
+      (_, tx) <- createAndStoreContract(
+        submittingParties = Set(alice),
+        signatories = Set(alice, bob),
+        stakeholders = Set(alice, bob, charlie),
+        key = None,
+      )
+      contractId = nonTransient(tx).loneElement
+      result <- ledgerDao.lookupActiveOrDivulgedContract(contractId, Set(charlie, emma))
+    } yield {
+      result.value shouldBe a[ContractInst[_]]
+    }
+  }
+
+  it should "find contract if at least one of requesters is a divulgee" in {
+    for {
+      (_, tx) <- createAndStoreContract(
+        submittingParties = Set(alice),
+        signatories = Set(alice, bob),
+        stakeholders = Set(alice, bob, charlie),
+        key = None,
+      )
+      contractId = nonTransient(tx).loneElement
+      _ <- storeCommitedContractDivulgence(
+        id = contractId,
+        divulgees = Set(emma)
+      )
+      result <- ledgerDao.lookupActiveOrDivulgedContract(contractId, Set(david, emma))
+    } yield {
+      result.value shouldBe a[ContractInst[_]]
+    }
+  }
+
+  it should "not find keys if none of requesters are stakeholders" in {
+    val aTextValue = ValueText(scala.util.Random.nextString(10))
+    for {
+      (_, _) <- createAndStoreContract(
+        submittingParties = Set(alice),
+        signatories = Set(alice, bob),
+        stakeholders = Set(alice, bob),
+        key = Some(KeyWithMaintainers(aTextValue, Set(alice, bob))),
+      )
+      key = GlobalKey.assertBuild(someTemplateId, aTextValue)
+      result <- ledgerDao.lookupKey(key, Set(charlie, emma))
+    } yield {
+      result shouldBe None
+    }
+  }
+
+  it should "find a key if at least one of requesters is a stakeholder" in {
+    val aTextValue = ValueText(scala.util.Random.nextString(10))
+    for {
+      (_, tx) <- createAndStoreContract(
+        submittingParties = Set(alice),
+        signatories = Set(alice, bob),
+        stakeholders = Set(alice, bob, charlie),
+        key = Some(KeyWithMaintainers(aTextValue, Set(alice, bob))),
+      )
+      contractId = nonTransient(tx).loneElement
+      key = GlobalKey.assertBuild(someTemplateId, aTextValue)
+      result <- ledgerDao.lookupKey(key, Set(emma, charlie))
+    } yield {
+      result.value shouldBe contractId
+    }
+  }
+
+  it should "not find a key if none of requesters are neither stakeholder nor divulgee" in {
```

Comment:
This test also overlaps a bit with "not find keys if none of requesters are stakeholders", where all requesters are unrelated parties.

The difference in this test is that one of the requesters is a divulgee, which is not enough to give contract key visibility. I would rename this test to "not find a key if the requesters are only divulgees" and set `divulgees = Set(david, emma)`.